### PR TITLE
devops: deployment pipeline

### PR DESCRIPTION
**Pull request by ternakkode**
The implementation of the AWS infrastructure might seem complex and overkill if it's seen as a solution for just deploying a Layer8 Server. However, some of the implementations here serve as a backbone for any other new product or microservice considering AWS. The implementation offers best practices, secure, cost-efficient, and scalable solutions. When introducing a new service, only a small implementation is required (you'll see no more than 4 new files for a new service to be integrated here).

Deployment Architecture TLDR:

- All AWS resources provisioned via Terraform, Terraform state stored in the Terraform cloud
- Layer8 Server running inside an EC2 Spot Machine, the deployment is managed via ECS (a simplified version of Kubernetes provided by AWS) with 1 GB RAM and 2 Core vCPU
- InfluxDB running inside an EC2 On Demand Machine with 1GB RAM and 2 Core CPU
- DNS and reverse proxy managed by Cloudflare
- InfluxDB and Layer8 Server instances running inside a private subnet without any public IP address. Both instances access the internet via the NAT Instance

Some resources to help with code review : 
- [Terraform Overview](https://www.youtube.com/watch?v=tomUWcQ0P3k&t=35s) to understand basic concepts about terraform and why terraform is a good option.)
- [Detailed video about how terraform work](https://www.youtube.com/watch?v=l5k1ai_GBDE&t=7s) to understand more details about Terraform.
- [Terraform Overview & Syntax Example](https://www.youtube.com/watch?v=HmxkYNv1ksg&t=290s) to understand terraform syntax
- [Article 1](https://www.hashicorp.com/resources/why-consider-terraform-enterprise-over-open-source),  [Article 2](https://medium.com/cloud-native-daily/comparing-terraform-cloud-and-open-source-terraform-for-infrastructure-as-code-27e39d4a5f8d) to understand why terraform cloud is chosen rather than fully using the open source solution
- [ECS Overview from AWS](https://www.youtube.com/watch?v=FnFvpIsBrog) to understand why we should use ECS to manage our layer8 deployment rather than just directly use the EC2 Instance
- [Detailed video about AWS ECS and comparison with Kubernetes](https://www.youtube.com/watch?v=AYAh6YDXuho) to understand more detail about AWS ECS infrastructure
- [Explanation about EC2 Spot 1](https://www.cloudforecast.io/blog/are-aws-spot-instances-worth-it-in-production/)  [Explanation about EC2 Spot 2](https://www.cloudzero.com/blog/spot-instances/) to understand why EC2 SPOT is the best option for Layer8 Server but not for InfluxDB 
- [EC2 Spot Overview video from AWS](https://www.youtube.com/watch?v=aRlY3VBX3Lc) to understand EC2 Spot basic concept
- [Explanation about Cloudflare Reverse Proxy](https://www.cloudflare.com/products/tunnel/) to understand why we use it
- [Basic Concept about how we do networking in AWS](https://www.uturndata.com/2021/02/23/aws-quick-tips-internet-gateways-nat-gateways-and-nat-instances/) to understand the concept of private subnet, public subnet, and NAT
- [Discussion and controversy about why we should have a private subnet](https://stackoverflow.com/questions/22188444/why-do-we-need-private-subnet-in-vpc) to understand why we should put our application in the private subnet besides starting February 2024 AWS costing 2.5$ per month for each public IP address.
- [Discussion started by a user who's annoyed about the high cost of NAT Gateway.](https://www.reddit.com/r/aws/comments/w3zrwz/nat_gateways_are_too_expensive/) to understand why we use NAT Instance rather than NAT Gateway

Cost Estimation per Environment : 
- NAT Instance: 5$
- NAT Public IP Address: 2.5$
- Layer8 Server = 2.30400 (EC2 Spot t3.micro ) + 2.85 (30GB SSD) = 5.1$
- InfluxDB Instance: 7.5 (EC On-Demand t3.micro) + 0.9 (10GB SSD) =  8.4$

Total : 21$ monthly exclude data transfer cost, S3 Storage and ECR storage

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -23,12 +23,12 @@ WORKDIR /layer8-app
 
 COPY --from=builder /build/main /layer8-app
 
-COPY --from=builder /build/.env /layer8-app
-
 COPY --from=builder /build/assets-v1 /layer8-app/assets-v1
 
 COPY --from=builder /build/certificates /layer8-app/certificates
 
+RUN touch /layer8-app/.env
```
Comment (ternakkode) [inline Dockerfile:30]: When we deploy the image to a service such as ECS/Kubernetes, we no longer use a .env file; instead, we store all configurations in the container environment variables. In this line of code, we add an empty .env file to ensure the application can still run locally, which is easier than putting the configuration in the environment variables.

File: .github/workflows/dev-service.yml [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+      - uses: actions/checkout@v4
+      - uses: kciter/aws-ecr-action@v3
+        with:
+          access_key_id: ${{ secrets.AWS_ACCESS_KEY_ID }}
+          secret_access_key: ${{ secrets.AWS_SECRET_ACCESS_KEY }}
+          account_id: ${{ secrets.AWS_ACCOUNT_ID }}
+          repo: layer8
+          region: ${{ vars.AWS_REGION }}
+          tags: latest,${{ github.sha }}
+          create_repo: true
```
Comment (ternakkode) [inline .github/workflows/dev-service.yml:23]: build docker image and push it to image registry, in this case we use AWS ECR

File: .github/workflows/dev-service.yml [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+          sudo apt-get install -y awscli
+      - name: Populate Application Environment Variables to S3
+        id: populateappenv
+        env:
+          AWS_ACCESS_KEY_ID: ${{ secrets.AWS_ACCESS_KEY_ID }}
+          AWS_SECRET_ACCESS_KEY: ${{ secrets.AWS_SECRET_ACCESS_KEY }}
+          AWS_DEFAULT_REGION: ${{ vars.AWS_REGION }}
+        run: |
+          echo -e "${{ vars.DEVELOPMENT_APP_ENV }}" >> app.env
+          aws s3 cp ./app.env s3://${{ vars.DEVELOPMENT_S3_APP_ENV_LOCATION }}
```
Comment (ternakkode) [inline .github/workflows/dev-service.yml:51]: In this code, we add all configuration or environment variable needed to run the layer8 server to the S3, we are using this to simplify the input of environment variable in the ECS task definition. The ECS will read all value from the S3 then put it to container environment variable automatically

File: .github/workflows/dev-service.yml [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+          echo -e "$TFVARS_STRING" >> terraform.tfvars
+        env:
+          TFVARS_STRING: |
+            ecr_image_tag = "${{ github.sha }}"
+            s3_arn_env_file = "arn:aws:s3:::${{ vars.DEVELOPMENT_S3_APP_ENV_LOCATION }}"
+            ecr_repository_url = "${{ vars.ECR_REPOSITORY_URL }}"
+            influxdb_url = "${{ vars.DEVELOPMENT_INFLUXDB_URL }}"
+            influxdb_token = "${{ secrets.DEVELOPMENT_INFLUXDB_TOKEN }}"
+            task_role_arn = "${{ secrets.ECS_TASK_ROLE_ARN }}"
+            task_execution_role_arn = "${{ secrets.ECS_TASK_EXECUTION_ROLE_ARN }}"
```
Comment (ternakkode) [inline .github/workflows/dev-service.yml:65]: ENV Variable needed as terraform input variable

File: .github/workflows/dev-influxdb.yml [hunk 5]
```diff
@@ -0,0 +1,39 @@
+
```
Comment (ternakkode) [inline .github/workflows/dev-influxdb.yml:1]: Basically, it's just a simple script to run Terraform in linux.

File: .github/workflows/dev-network.yml [hunk 6]
```diff
@@ -0,0 +1,32 @@
+name: Deploy Network to Production Environment
```
Comment (ternakkode) [inline .github/workflows/dev-network.yml:1]: Basically, it's just a simple script to run Terraform in linux.

File: .github/workflows/prod-influxdb.yml [hunk 7]
```diff
@@ -0,0 +1,39 @@
+
```
Comment (ternakkode) [inline .github/workflows/prod-influxdb.yml:1]: Basically, it's just a simple script to run Terraform in linux.

File: .github/workflows/prod-network.yml [hunk 8]
```diff
@@ -0,0 +1,32 @@
+name: Deploy Network to Production Environment
```
Comment (ternakkode) [inline .github/workflows/prod-network.yml:1]: Basically, it's just a simple script to run Terraform in linux.

File: .github/workflows/prod-service.yml [hunk 9]
```diff
@@ -0,0 +1,75 @@
+name: Deploy Layer8 Server to Production Environment
```
Comment (ternakkode) [inline .github/workflows/prod-service.yml:1]: same as how we do for dev environment

File: infra/production/base/main.tf [hunk 10]
```diff
@@ -0,0 +1,40 @@
+terraform {
```
Comment (ternakkode) [inline infra/production/base/main.tf:1]: Define all dependencies needed to run Layer8, InfluxDB, and any new services that may be introduced in the future. This base folder should be moved to a new repository when our other products start using AWS.

In this code, we define:

- Secure and cost-efficient networking solution
- IAM Role to access AWS ECS
- S3 bucket for storing ECS container config
- ECR for storing docker image

File: infra/production/base/module/ecr/main.tf [hunk 11]
```diff
@@ -0,0 +1,8 @@
+resource "aws_ecr_repository" "layer8_ecr_repo" {
+  name                 = "layer8"
+  image_tag_mutability = "MUTABLE"
+
+  image_scanning_configuration {
+    scan_on_push = true
+  }
+}
```
Comment (ternakkode) [inline infra/production/base/module/ecr/main.tf:8]: Create an ECR Repository for storing Docker Image

File: infra/production/base/module/ecs/capacity_provider.tf [hunk 12]
```diff
@@ -0,0 +1,62 @@
+resource "aws_launch_template" "service_spot_launch_template" {
```
Comment (ternakkode) [inline infra/production/base/module/ecs/capacity_provider.tf:1]: TLDR: This file essentially creates an AWS configuration that automatically generates an AWS EC2 SPOT Instance with specified configurations, alleviating concerns about EC2 Spot Interruptions while still benefiting from cost-effective pricing.

File: infra/production/base/module/ecs/main.tf [hunk 13]
```diff
@@ -0,0 +1,44 @@
+resource "aws_ecs_cluster" "cluster" {
```
Comment (ternakkode) [inline infra/production/base/module/ecs/main.tf:1]: TLDR: This file defines the ECS cluster, IP Whitelist for ECS Instances, and specifies where the ECS can obtain an EC2 instance to run the container.

File: infra/production/base/module/iam/main.tf [hunk 14]
```diff
@@ -0,0 +1,67 @@
+data "aws_iam_policy_document" "ecs_node_doc" {
```
Comment (ternakkode) [inline infra/production/base/module/iam/main.tf:1]: TLDR: This file is used to define an AWS IAM Role to grant AWS ECS access to our AWS infrastructure.

File: infra/production/base/module/network/main.tf [hunk 15]
```diff
@@ -0,0 +1,93 @@
+data "aws_availability_zones" "available" { state = "available" }
```
Comment (ternakkode) [inline infra/production/base/module/network/main.tf:1]: TLDR: Define VPC with private and public subnet topology and use NAT Instance from open source solution

File: infra/production/influxdb/main.tf [hunk 16]
```diff
@@ -0,0 +1,59 @@
+terraform {
```
Comment (ternakkode) [inline infra/production/influxdb/main.tf:1]: TLDR: Create an EC2 instance for InfluxDB and execute a script to install InfluxDB and the Cloudflare tunnel inside the EC2 instance.

File: infra/production/layer8/main.tf [hunk 17]
```diff
@@ -0,0 +1,179 @@
...
+
+  config = {
+    organization = "globe-and-citizen"
+    workspaces = {
+      name = "network-production"
+    }
+  }
+}
+
+resource "aws_ecs_task_definition" "task_definition" {
```
Comment (ternakkode) [inline infra/production/layer8/main.tf:25]: TLDR: task definition is almost similar as docker compose

File: infra/production/layer8/main.tf [hunk 18]
```diff
@@ -0,0 +1,179 @@
...
+  family                   = "layer8-server"
+  task_role_arn            = data.terraform_remote_state.network.outputs.task_role_arn
+  execution_role_arn       = data.terraform_remote_state.network.outputs.task_execution_role_arn
+  network_mode             = "awsvpc"
+  requires_compatibilities = ["EC2"]
+  skip_destroy             = true
+
+  container_definitions = jsonencode([
+    {
```
Comment (ternakkode) [inline infra/production/layer8/main.tf:34]: define container for layer8 server

File: infra/production/layer8/main.tf [hunk 19]
```diff
@@ -0,0 +1,179 @@
...
+      portMappings = [
+        { containerPort = 5001, hostPort = 5001, protocol = "tcp" },
+      ],
+      environment = [],
+      environmentFiles = [
+        { type = "s3", value = var.s3_arn_env_file }
+      ],
+      systemControls = [],
+      volumesFrom    = [],
+      logConfiguration = {
```
Comment (ternakkode) [inline infra/production/layer8/main.tf:50]: define how we store layer8 server logs

File: infra/production/layer8/main.tf [hunk 20]
```diff
@@ -0,0 +1,179 @@
...
+          "awslogs-create-group" : "true",
+          "awslogs-group" : "ecs/production",
+          "awslogs-region" : "${var.aws_region}",
+          "awslogs-stream-prefix" : "layer8server"
+        },
+      },
+      user = "0"
+    },
+    {
+      name              = "cloudflared-tunnel",
```
Comment (ternakkode) [inline infra/production/layer8/main.tf:62]: define cloudflare tunnel as reverse proxy

File: infra/production/layer8/main.tf [hunk 21]
```diff
@@ -0,0 +1,179 @@
...
+      command = [
+        "tunnel",
+        "--no-autoupdate",
+        "run",
+        "--token",
+        "${var.cloudflare_tunnel_token}"
+      ]
+    },
+    {
+      name              = "telegraf",
```
Comment (ternakkode) [inline infra/production/layer8/main.tf:92]: define telegraf as a sidecar to forward metrics from layer8 server to influxdb

File: infra/production/layer8/main.tf [hunk 22]
```diff
@@ -0,0 +1,179 @@
...
+      command = []
+    }
+  ])
+
+  tags = {
+    Name = "layer8-server"
+  }
+}
+
+resource "aws_ecs_service" "service" {
```
Comment (ternakkode) [inline infra/production/layer8/main.tf:139]: define the runtime configuration for the layer8 server

File: infra/production/layer8/main.tf [hunk 23]
```diff
@@ -0,0 +1,179 @@
...
+  name            = "layer8-server"
+  cluster         = data.terraform_remote_state.network.outputs.ecs_cluster_id
+  task_definition = aws_ecs_task_definition.task_definition.arn
+  desired_count   = 1
```
Comment (ternakkode) [inline infra/production/layer8/main.tf:143]: define how the service horizontally scaled

**Review by ternakkode — COMMENTED** [review]
Comments explaining the implementation

File: infra/development/base/main.tf [hunk 24]
```diff
@@ -0,0 +1,40 @@
+terraform {
```
Comment (ternakkode) [inline infra/development/base/main.tf:1]: Skip reviewing any file in the infra/development directory, as some of it duplicates code in the infra/production folder. I've added comments to explain the implementation.

File: infra/production/influxdb/main.tf [hunk 25]
```diff
@@ -0,0 +1,59 @@
...
+    }
+  }
+}
+
+resource "aws_instance" "ec2_instance" {
+  tags = {
+    Name = "influxdb2-production"
+  }
+
+  ami           = "ami-023adaba598e661ac"
```
Comment (ternakkode) [inline infra/production/influxdb/main.tf:30]: You can see this as the ID of the Ubuntu 2022 OS stored in AWS.

File: infra/production/layer8/main.tf [hunk 26]
```diff
@@ -0,0 +1,179 @@
...
+data "terraform_remote_state" "network" {
+  backend = "remote"
+
+  config = {
+    organization = "globe-and-citizen"
+    workspaces = {
+      name = "network-production"
+    }
+  }
+}
```
Comment (ternakkode) [inline infra/production/layer8/main.tf:23]: read data from other terraform workspace

File: infra/production/layer8/main.tf [hunk 27]
```diff
@@ -0,0 +1,179 @@
...
+  capacity_provider_strategy {
+    capacity_provider = data.terraform_remote_state.network.outputs.service_spot_capacity_provider_name
+    base              = 1
+    weight            = 100
+  }
+
+  ordered_placement_strategy {
+    type  = "spread"
+    field = "attribute:ecs.availability-zone"
+  }
```
Comment (ternakkode) [inline infra/production/layer8/main.tf:166]: define the source of ec2 instance to run the container

**Comment by huzaifamk** [comment]
Great expalanation @ternakkode , well done!

**Review by huzaifamk — APPROVED** [review]
I am done reviewing and learning from this, great explanation from @ternakkode.

@NJavokhir if you also review, you will also learn, please add your review approving the PR once you are done, then I will get this merged.

**Comment by ternakkode** [comment]
> I am done reviewing and learning from this, great explanation from @ternakkode.
> 
> @NJavokhir if you also review, you will also learn, please add your review approving the PR once you are done, then I will get this merged.



> I am done reviewing and learning from this, great explanation from @ternakkode.
> 
> @NJavokhir if you also review, you will also learn, please add your review approving the PR once you are done, then I will get this merged.

Let's get this merged first to unblock our progress. Later, Javokhir still can learn about these changes.

**Pull request merged**